nixos/modules/virtualisation/qemu-vm.nix

```diff
@@ -598,17 +602,34 @@ in
 
     # After booting, register the closure of the paths in
     # `virtualisation.pathsInNixDB' in the Nix database in the VM.  This
-    # allows Nix operations to work in the VM.  The path to the
-    # registration file is passed through the kernel command line to
-    # allow `system.build.toplevel' to be included.  (If we had a direct
-    # reference to ${regInfo} here, then we would get a cyclic
-    # dependency.)
-    boot.postBootCommands =
+    # allows Nix operations to work in the VM.
+    # When running without a bootloader, the path to the registration file is
+    # passed through the kernel command line to allow `system.build.toplevel'
+    # to be included.  (If we had a direct reference to ${regInfo} here, then
+    # we would get a cyclic dependency.)
+    # When running with a bootloader, we instead pass through a file in /boot,
+    # which is loaded by systemd after the fs is mounted. (This is to support
+    # both systemd-boot and grub.)
+    boot.postBootCommands = if !cfg.useBootLoader then
       ''
         if [[ "$(cat /proc/cmdline)" =~ regInfo=([^ ]*) ]]; then
           ${config.nix.package.out}/bin/nix-store --load-db < ''${BASH_REMATCH[1]}
         fi
-      '';
+      ''
+      else "";
+
+    systemd.services = if cfg.useBootLoader then {
+      nix-load-db = {
+        description = "Load nix db shared by host";
+        wantedBy = [ "default.target" ];
+        wants = ["boot.mount"];
+        after = ["boot.mount"];
```

Comment:
It may be nice to use a [RequiresMountsFor](https://www.freedesktop.org/software/systemd/man/systemd.unit.html#RequiresMountsFor=), but this doesn't seem to be exposed by [systemd.nix](https://github.com/NixOS/nixpkgs/blob/nixos-21.05/nixos/modules/system/boot/systemd.nix)